### DISCUSSION TITLE
How evaluate performance on Planning engine

### DISCUSSION THREAD

**Discussion by sgfisme** [frePPLe/frepple#508] [Q&A]
We make a FrePPLe plan, checked the FrePPLe log and separated the execution time into 3 part (data init, engine execution and export plan).
The plan is very complex and the total execution time is 244 minutes(data init 7 minutes, engine execution 55 minutes and export plan 183 minutes).
In addition to improve the performance of export plan, we also desire to improve the engine execution performance.

Could anybody give some tuning advises about frePPLe ?

1. How to identify the FrePPLe engine detail execution step and time? Adjust log level or ... ?
2. How to improve the engine execution performance ? Adjust system parameter, or ...

**Comment by jdetaeye** [comment]
>>How to identify the FrePPLe engine detail execution step and time? Adjust log level or ... ?

For production runs you definitely want to run with log level 0.   All other debugging levels disable multi-threaded execution - see previous discussions on this topic. 

To understand where the planning engine spends its time, you run with loglevel 2 and review the how many steps the algorithm needs to plan some example demands. In a lot of cases it allows you to spot situations where the search algorithm goes inefficient (ie many thousands of lines in the log file to plan a single demand) and you can adjust your model or some algorithm parameters. 

As a last resort for speeding things up and if the search algorithm goes normal and logical, you can compile a profiling version of the planning engine and run your dataset with it.  The output shows which parts of the c++ code are critical in your model and potential areas for optimization.  This is a complex analysis that requires in depth understanding of the code and algorithm structure.

**Reply by sgfisme** [reply]
> For production runs you definitely want to run with log level 0. All other debugging levels disable multi-threaded execution - see previous discussions on this topic.
> 

How to calculate how many threads use the engine? Does it have limit? Or it calculate based on the planning algorithm ?
